ccip.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/wealdtech/go-ens/v3/contracts/offchainresolver"
+	"github.com/wealdtech/go-ens/v3/contracts/universalresolver"
+)
+
+func getCcipReadError(err error) (bool, string) {
+	var jsonErr, ok = err.(rpc.DataError)
+	if !ok {
+		return false, ""
+	}
+	errData, ok := jsonErr.ErrorData().(string)
+	if !ok {
+		return false, ""
+	}
+	return (len(errData) >= 10 && errData[:10] == offchainLookupSignature), errData
+}
+
+func CCIPRead(backend bind.ContractBackend, rAddr common.Address, revertData string) ([]byte, error) {
+	hexBytes, err := hex.DecodeString(revertData[2:])
+	if err != nil {
+		return nil, err
+	}
+
+	uAbi, err := universalresolver.ContractMetaData.GetAbi()
+	if err != nil {
+		return nil, err
+	}
+
+	// Extracting error details from the revert data
+	var sig [4]byte
+	copy(sig[:], hexBytes[:4])
+	abiErr, err := uAbi.ErrorByID(sig)
+	if err != nil {
+		return nil, err
+	}
+	errArgs, err := abiErr.Inputs.Unpack(hexBytes[4:])
+	if err != nil {
+		return nil, err
+	}
+
+	sender := errArgs[0].(common.Address)
+	urls := errArgs[1].([]string)
+	calldata := common.Bytes2Hex(errArgs[2].([]byte))
+	calldataHex := fmt.Sprintf("0x%s", calldata)
+	callback := errArgs[3].([4]byte)
+	extraData := common.Bytes2Hex(errArgs[4].([]byte))
+	extraDataHex := fmt.Sprintf("0x%s", extraData)
+
+	// Fetching data from external source using CCIP
+	resp, err := CCIPFetch(sender, calldataHex, urls)
+	if err != nil || len(resp) == 0 {
+		return nil, errors.New("unregistered name")
+	}
+
+	oAbi, err := offchainresolver.ContractMetaData.GetAbi()
+	if err != nil {
+		return nil, errors.New("no address")
+	}
+	m, err := oAbi.MethodById(callback[:])
+	if err != nil {
+		return nil, errors.New("no address")
+	}
+	args, err := m.Inputs.Pack(common.FromHex(resp), common.FromHex(extraDataHex))
+	if err != nil {
+		return nil, errors.New("no address")
+	}
+
+	encodedResp, err := backend.CallContract(context.Background(), ethereum.CallMsg{
+		To:   &rAddr,
+		Data: append(callback[:], args...),
+	}, nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	outputs, err := m.Outputs.Unpack(encodedResp)
+	if err != nil || len(outputs) == 0 {
+		return nil, err
+	}
+	return outputs[0].([]byte), nil
+}
+
+func CCIPFetch(sender common.Address, data string, urls []string) (result string, err error) {
+	for _, url := range urls {
+		method := "POST"
+		if strings.Contains(url, "{data}") {
+			method = "GET"
+		}
+
+		body := []byte{}
+		if method == "POST" {
+			body, err = json.Marshal(map[string]interface{}{
+				"data":   data,
+				"sender": sender,
+			})
+			if err != nil {
+				return "", err
+			}
+		}
+
+		url = strings.ReplaceAll(url, "{sender}", sender.String())
+		url = strings.ReplaceAll(url, "{data}", data)
+		req, err := http.NewRequest(method, url, bytes.NewBuffer(body))
+		if err != nil {
+			return "", err
+		}
+		req.Header.Set("Content-Type", "application/json")
+
+		client := &http.Client{}
```

Comment:
There should be a timeout here to avoid the request hanging.

Reply:
thoughts on the user of the library passing in an http client vs creating a new client for every request?